aggregator/src/aggregator/accumulator.rs

```diff
@@ -137,7 +137,22 @@ impl<const SEED_SIZE: usize, Q: AccumulableQueryType, A: vdaf::Aggregator<SEED_S
     ) -> Result<HashSet<ReportId>, datastore::Error> {
         let unmergeable_report_ids = Arc::new(Mutex::new(HashSet::new()));
 
-        try_join_all(self.aggregations.values().map(|data| {
+        // Prevent deadlocks when inserting into shards by providing a deterministic order to shard
+        // updates. Suppose two concurrent processes are attempting to update a batch aggregation.
+        // We want to avert this situation:
+        //
+        // A> BEGIN;
+        // A> UPDATE batch_aggregations WHERE ord = 1 ... -- Row with ord 1 is locked for update.
+        // B> BEGIN;
+        // B> UPDATE batch_aggregations WHERE ord = 2 ... -- Row with ord 2 is locked for update.
+        // A> UPDATE batch_aggregations WHERE ord = 2 ... -- A is now blocked waiting for B to finish.
+        // B> UPDATE batch_aggregations WHERE ord = 1 ... -- Kaboom!
+        //
+        // This situation cannot happen if the UPDATE statements are ordered by `ord`.
+        let mut aggregations: Vec<_> = self.aggregations.values().collect();
+        aggregations.sort_unstable_by_key(|data| data.batch_aggregation.ord());
+
+        try_join_all(aggregations.into_iter().map(|data| {
```

Comment:
N.B. this isn't a total solution because [`try_join_all`](https://docs.rs/futures/latest/futures/future/fn.try_join_all.html) drives its futures concurrently, with no guaranteed order. Since we see a performance improvement, it seems that the current implementation will (usually?) drive the futures in the given order at least to the point that a deadlock would be created. If the implementation changes in a futures update, we might see a regression here.

All that said, a performance win is a performance win. I have no objection to merging this as-is, since the apparent "solution" to enforce ordering would be to run these transactions serially -- and that would be a big perf loss.

Reply:
Good point! I'll fix my comment to point this out.